tests/test_controlsocket.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# Licensed under the GPLv3, see LICENSE file for details.
+
+import io
+import unittest
+import urllib.error
+from controlsocket import Client, APIError, ConnectionError
+
+
+class TestClass(unittest.TestCase):
+    def test_add_metrics_user_success(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body=r'{"username": "juju-metrics-r0", "password": "passwd"}',
+
+            response=mockResponse(
+                headers=mockHeaders(content_type='application/json'),
+                body=r'{"message":"created user \"juju-metrics-r0\""}'
+            )
+        )
+        control_socket.add_metrics_user('juju-metrics-r0', 'passwd')
+
+    def test_add_metrics_user_fail(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body=r'{"username": "juju-metrics-r0", "password": "passwd"}',
+
+            error=urllib.error.HTTPError(
+                url='http://localhost/metrics-users',
+                code=409,
+                msg='',
+                hdrs=None,
+                fp=io.BytesIO(br'{"error":"user \"juju-metrics-r0\" already exists"}'),
+            )
+        )
+
+        self.assertRaisesRegex(
```

Comment:
You'd normally use the `with` statement with `assertRaises*` so that you can make a regular method call and the code reads cleaner, like so:

```python
with self.assertRaisesRegex(APIError, r'regex'):
    control_socket.add_metrics_user('juju-metrics-r0', 'passwd')
```

I think we should also test that the APIError `.body`, `.code`, `.status`, and `.message` attributes are populated correctly. Something like this ([see docs](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises)):

```python
with self.assertRaises(APIError) as cm:
    control_socket.add_metrics_user('juju-metrics-r0', 'passwd')
self.assertEqual(cm.exception.code, 409)
self.assertEqual(cm.exception.status, 'blah')
self.assertEqual(cm.exception.message, 'blah blah')
```